polygeist/lib/Dialect/Polygeist/Utils/TransformUtils.cpp

```diff
@@ -399,3 +402,62 @@ void LoopTools::versionLoop(LoopLikeOpInterface loop,
                             const LoopVersionCondition &versionCond) const {
   LoopVersionBuilder(loop).versionLoop(versionCond);
 }
+
+//===----------------------------------------------------------------------===//
+// SYCL related functions
+//===----------------------------------------------------------------------===//
+
+static constexpr StringRef linkageAttrName = "llvm.linkage";
+bool mlir::isLinkonceODR(FunctionOpInterface func) {
+  if (!func->hasAttr(linkageAttrName))
+    return false;
+  auto attr = dyn_cast<LLVM::LinkageAttr>(func->getAttr(linkageAttrName));
+  assert(attr && "Expecting LLVM::LinkageAttr");
+  return attr.getLinkage() == LLVM::Linkage::LinkonceODR;
+}
+
+bool mlir::isOnlyCalledFromGPUKernel(FunctionOpInterface func) {
+  ModuleOp module = func->getParentOfType<ModuleOp>();
+  SymbolTableCollection symTable;
+  SymbolUserMap userMap(symTable, module);
+  return all_of(userMap.getUsers(func), [&](Operation *call) {
+    if (auto callerFunc = call->getParentOfType<FunctionOpInterface>()) {
+      Operation *op = callerFunc;
+      if (auto gpuFunc = dyn_cast<gpu::GPUFuncOp>(op))
+        if (gpuFunc.isKernel())
+          return true;
+      return isOnlyCalledFromGPUKernel(cast<FunctionOpInterface>(callerFunc));
+    }
+    return false;
+  });
+}
+
+bool mlir::isPotentialKernelBodyFunc(FunctionOpInterface func) {
+  // The function must be defined, and private or with linkonce_odr linkage.
+  if (func.isExternal() || (!func.isPrivate() && !isLinkonceODR(func)))
+    return false;
+
+  ModuleOp module = func->getParentOfType<ModuleOp>();
+  SymbolTableCollection symTable;
+  SymbolUserMap userMap(symTable, module);
+
+  // Ensure all the call sites for this function are tail calls.
+  if (any_of(userMap.getUsers(func), [](Operation *call) {
+        if (!call->getBlock()->hasNoSuccessors())
+          return true;
+        Operation *op = call->getNextNode();
+        while (op) {
+          if (!isMemoryEffectFree(op))
+            return true;
+          op = op->getNextNode();
+        }
+        return false;
```

Comment:
Can you make this code a utility function named `isTailCall` that takes a CallOpInterface ?